### PR TITLE
feat(cli): add cache clean command

### LINKED ISSUE #1127: Add cache clean subcommand
### Component

Forge

### Describe the feature you would like

forked state is now being cache. to remove this data there should be a simple `forge cache clean <chain(s)|all> <block(s)|all>` subcommand that removes the cache files.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by adaddeo**
❗extending this command with options to specify chain and blocks. first time working in Rust, feedback would be appreciated.

## Motivation
closes #1127 

forked state is now being added to the cache. this state should be easily cleanable.

## Solution
adding a cli command `forge cache clean` to clean the cache.

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
thanks so much,
looking good so far,

what we'd need in the `CleanArgs` are additional options for chain and block, ideally multiple args, so that `forge cache clean all`, or `forge cache clean mainnet fantom optimism` works

**Comment by adaddeo** [comment]
> thanks so much, looking good so far,
> 
> what we'd need in the `CleanArgs` are additional options for chain and block, ideally multiple args, so that `forge cache clean all`, or `forge cache clean mainnet fantom optimism` works

got it. thinking abt how multiple chains + multiple blocks could be specified, I came up w/ the following:
- if no chain or "all" is supplied, blocks can not be specified
- if one chain is supplied, multiple blocks can be specified
- if multiple chains are supplied, blocks can not be specified

sounds reasonable? other ideas? also would be in favor of scoping this down to the simplest acceptable use case worth merging and opening a new issue for the extended functionality (first time using rust so might take me a min to find the time to read clap docs and setup more complex args)

**Comment by mattsse** [comment]
great!

> if one chain is supplied, multiple blocks can be specified
> if multiple chains are supplied, blocks can not be specified

I think we could combine these two so that `forge cache clean mainnet fantom optimism --blocks 100 200` is also supported but this will probably not used very often so we don't really need it I'd asy

File: cli/src/cmd/forge/cache.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            "polygon",
+            "polygon_mumbai",
+            "avalanche",
+            "avalanche_fuji",
+            "sepolia",
+            "moonbeam",
+            "moonbeam_dev",
+            "moonriver",
+            "optimism",
+            "optimism-kovan"
```
Comment (mattsse) [inline cli/src/cmd/forge/cache.rs:61]: I think this is fine for now, wonder if clap allows this to be a function instead

File: cli/tests/it/cmd.rs [hunk 2]
```diff
@@ -30,6 +30,13 @@ forgetest!(can_clean_non_existing, |prj: TestProject, mut cmd: TestCommand| {
     prj.assert_cleaned();
 });
 
+// checks that `cache clean` can be invoked and cleans the foundry cache
+forgetest!(can_cache_clean, |prj: TestProject, mut cmd: TestCommand| {
+    cmd.args(["cache", "clean"]);
+    cmd.assert_empty_stdout();
+    prj.assert_foundry_cache_cleaned();
+});
+
```
Comment (mattsse) [inline cli/tests/it/cmd.rs:39]: great!
could we also mock some empty block files here?
just creating empty files for a chain and test with chain and block argument as well?

File: config/src/lib.rs [hunk 3]
```diff
@@ -784,21 +791,57 @@ impl Config {
...
-                }
+                };
             }
             let cwd = std::env::current_dir().ok()?;
             let mut cwd = cwd.as_path();
             loop {
                 let file_path = cwd.join(path);
                 if file_path.is_file() {
-                    return Some(file_path)
+                    return Some(file_path);
```
Comment (mattsse) [inline config/src/lib.rs:801]: fyi we use nightly formatting `cargo +nightly fmt`

**Review by mattsse — CHANGES_REQUESTED** [review]
this is looking great!
only smol nit

File: config/src/lib.rs (hunk 3 above)
Comment (adaddeo) [reply config/src/lib.rs]: removed fmt patches from commit

File: cli/tests/it/cmd.rs (hunk 2 above)
Comment (adaddeo) [reply cli/tests/it/cmd.rs]: added tests for `cache clean <chain>` and `cache clean <chain> --block <blocks>`. in each test I'm setting up directories at the start of the test so our assertions that the paths don't exist won't pass by default.

any preference or pointers on how to best isolate these tests? rn they operate on `~` and can fail if run in parallel. it looks like there are crates for mocking or working with a [fake file system](https://github.com/iredelmeier/filesystem-rs) but didn't see any of them being used by the project currently

**Review by mattsse — APPROVED** [review]
awesome!
this good to merge?

**Comment by mattsse** [comment]
oh I see, this would have the side-effect that running tests locally the cache gets wiped which is not ideal.
don't think it's worth making the cache path configurable, so given that this works, let's simply ignore these tests: `forgetest_ingore!` macro

if there are problems with the cache we can investigate.

**Comment by adaddeo** [comment]
@mattsse ignored tests

should be gtg

**Review by mattsse — APPROVED** [review]
thank you!

**Comment by mattsse** [comment]
also smol linting issue

**Pull request merged**